File: mycollab-services/src/main/java/com/esofthead/mycollab/module/crm/service/ibatis/QuoteServiceImpl.java
/**
 * This file is part of mycollab-services.
 *
 * mycollab-services is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-services is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-services.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.crm.service.ibatis;

import com.esofthead.mycollab.core.persistence.ICrudGenericDAO;
import com.esofthead.mycollab.core.persistence.ISearchableDAO;
import com.esofthead.mycollab.core.persistence.service.DefaultService;
import com.esofthead.mycollab.module.crm.dao.ProductMapper;
import com.esofthead.mycollab.module.crm.dao.QuoteMapper;
import com.esofthead.mycollab.module.crm.dao.QuoteMapperExt;
import com.esofthead.mycollab.module.crm.domain.*;
import com.esofthead.mycollab.module.crm.domain.criteria.QuoteSearchCriteria;
import com.esofthead.mycollab.module.crm.service.QuoteGroupProductService;
import com.esofthead.mycollab.module.crm.service.QuoteService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@Transactional
public class QuoteServiceImpl extends DefaultService<Integer, Quote, QuoteSearchCriteria> implements
        QuoteService {

    @Autowired
    private QuoteMapper quoteMapper;
    @Autowired
    private QuoteMapperExt quoteMapperExt;
    @Autowired
    private QuoteGroupProductService quoteGroupProductService;
    @Autowired
    private ProductMapper productMapper;

    @Override
    public ICrudGenericDAO<Integer, Quote> getCrudMapper() {
        return quoteMapper;
    }

    @Override
    public ISearchableDAO<QuoteSearchCriteria> getSearchMapper() {
        return quoteMapperExt;
    }

    public void setProductDAO(ProductMapper productDAO) {
        this.productMapper = productDAO;
    }

    @Override
    public void saveSimpleQuoteGroupProducts(int accountid, int quoteId,
                                             List<SimpleQuoteGroupProduct> entity) {
        quoteGroupProductService.deleteQuoteGroupByQuoteId(quoteId);

        for (SimpleQuoteGroupProduct simpleQuoteGroupProduct : entity) {
            QuoteGroupProduct quoteGroupProduct = simpleQuoteGroupProduct
                    .getQuoteGroupProduct();
            quoteGroupProductService.insertQuoteGroupExt(quoteGroupProduct);

            for (Product quoteProduct : simpleQuoteGroupProduct
                    .getQuoteProducts()) {
                // quoteProduct.setAccountid(accountid);
                quoteProduct.setGroupid(quoteGroupProduct.getId());
                quoteProduct.setStatus("Quoted");
                productMapper.insert(quoteProduct);
            }
        }
    }

    @Override
    public List<SimpleQuoteGroupProduct> getListSimpleQuoteGroupProducts(
            int quoteId) {
        List<QuoteGroupProduct> quoteGroupProducts = quoteGroupProductService
                .findQuoteGroupByQuoteId(quoteId);

        List<SimpleQuoteGroupProduct> result = new ArrayList<SimpleQuoteGroupProduct>();
        for (QuoteGroupProduct quoteGroupProduct : quoteGroupProducts) {
            SimpleQuoteGroupProduct simpleQuoteGroupProduct = new SimpleQuoteGroupProduct();
            simpleQuoteGroupProduct.setQuoteGroupProduct(quoteGroupProduct);

            ProductExample quoteProductEx = new ProductExample();
            quoteProductEx.createCriteria().andGroupidEqualTo(
                    quoteGroupProduct.getId());
            List<Product> quoteProducts = productMapper
                    .selectByExample(quoteProductEx);
            simpleQuoteGroupProduct.setQuoteProducts(quoteProducts);
            result.add(simpleQuoteGroupProduct);
        }
        return result;
    }
}

File: mycollab-services/src/main/java/com/mycollab/module/crm/service/impl/QuoteServiceImpl.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.crm.service.impl

import com.mycollab.db.persistence.ICrudGenericDAO
import com.mycollab.db.persistence.ISearchableDAO
import com.mycollab.db.persistence.service.DefaultService
import com.mycollab.module.crm.dao.ProductMapper
import com.mycollab.module.crm.dao.QuoteMapper
import com.mycollab.module.crm.dao.QuoteMapperExt
import com.mycollab.module.crm.domain.ProductExample
import com.mycollab.module.crm.domain.Quote
import com.mycollab.module.crm.domain.SimpleQuoteGroupProduct
import com.mycollab.module.crm.domain.criteria.QuoteSearchCriteria
import com.mycollab.module.crm.service.QuoteGroupProductService
import com.mycollab.module.crm.service.QuoteService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class QuoteServiceImpl(private val quoteMapper: QuoteMapper,
                       private val quoteMapperExt: QuoteMapperExt,
                       private val quoteGroupProductService: QuoteGroupProductService,
                       private var productMapper: ProductMapper) : DefaultService<Int, Quote, QuoteSearchCriteria>(), QuoteService {

    override val crudMapper: ICrudGenericDAO<Int, Quote>
        get() = quoteMapper as ICrudGenericDAO<Int, Quote>

    override val searchMapper: ISearchableDAO<QuoteSearchCriteria>
        get() = quoteMapperExt

    fun setProductDAO(productDAO: ProductMapper) {
        this.productMapper = productDAO
    }

    override fun saveSimpleQuoteGroupProducts(accountid: Int, quoteId: Int, entities: List<SimpleQuoteGroupProduct>) {
        quoteGroupProductService.deleteQuoteGroupByQuoteId(quoteId)

        for (simpleQuoteGroupProduct in entities) {
            val quoteGroupProduct = simpleQuoteGroupProduct.quoteGroupProduct
            quoteGroupProductService.insertQuoteGroupExt(quoteGroupProduct)

            for (quoteProduct in simpleQuoteGroupProduct.quoteProducts) {
                // quoteProduct.setAccountid(accountid);
                quoteProduct.groupid = quoteGroupProduct.id
                quoteProduct.status = "Quoted"
                productMapper.insert(quoteProduct)
            }
        }
    }

    override fun getListSimpleQuoteGroupProducts(quoteId: Int): List<SimpleQuoteGroupProduct> {
        val quoteGroupProducts = quoteGroupProductService.findQuoteGroupByQuoteId(quoteId)

        val result = ArrayList<SimpleQuoteGroupProduct>()
        for (quoteGroupProduct in quoteGroupProducts) {
            val simpleQuoteGroupProduct = SimpleQuoteGroupProduct()
            simpleQuoteGroupProduct.quoteGroupProduct = quoteGroupProduct

            val quoteProductEx = ProductExample()
            quoteProductEx.createCriteria().andGroupidEqualTo(quoteGroupProduct.id)
            val quoteProducts = productMapper.selectByExample(quoteProductEx)
            simpleQuoteGroupProduct.quoteProducts = quoteProducts
            result.add(simpleQuoteGroupProduct)
        }
        return result
    }
}
